shard.go

```diff
@@ -0,0 +1,133 @@
+package flatfs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+type ShardIdV1 struct {
+	funName string
+	param   int
+	fun     ShardFunc
+}
+
+const PREFIX = "/repo/flatfs/shard/"
+
+func (f *ShardIdV1) String() string {
+	return fmt.Sprintf("%sv1/%s/%d", PREFIX, f.funName, f.param)
+}
+
+func (f *ShardIdV1) Func() ShardFunc {
+	return f.fun
+}
+
+func ParseShardFunc(str string) (*ShardIdV1, error) {
+	str = strings.TrimSpace(str)
+	// ignore prefix for now
+	if len(str) == 0 {
+		return nil, fmt.Errorf("empty shard identifier")
+	}
+	if str[0] == '/' {
+		if !strings.HasPrefix(str, PREFIX) {
+			return nil, fmt.Errorf("invalid prefix in shard identifier: %s", str)
+		}
+		str = str[len(PREFIX):]
+	}
+	parts := strings.Split(str, "/")
+	if len(parts) == 3 {
+		version := parts[0]
+		if version != "v1" {
+			return nil, fmt.Errorf("expected 'v1' for version string got: %s\n", version)
+		}
+		parts = parts[1:]
+	}
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid shard identifier: %s", str)
+	}
+
+	id := &ShardIdV1{funName: parts[0]}
+
+	param, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return nil, fmt.Errorf("invalid parameter: %v", err)
+	}
+	id.param = param
+
+	switch id.funName {
+	case "prefix":
+		id.fun = Prefix(param)
+	case "suffix":
+		id.fun = Suffix(param)
+	case "next-to-last":
+		id.fun = NextToLast(param)
+	default:
+		return nil, fmt.Errorf("expected 'prefix', 'suffix' or 'next-to-last' got: %s", id.funName)
+	}
+
+	return id, nil
+}
+
+func ReadShardFunc(dir string) (*ShardIdV1, error) {
+	buf, err := ioutil.ReadFile(filepath.Join(dir, "SHARDING"))
```

Comment:
Fine, Done.  However, I think it makes code harder to read and I also am horrible at naming the things.